### ISSUE TITLE
DynamoDB DynamoDBDocument update will not accept objects (Maps) in JSON even if they are pre-marshalled

### ISSUE THREAD

**Issue by third-party-integrations-flyht** [aws/aws-sdk-js-v3#5487]
### Checkboxes for prior research

- [X] I've gone through [Developer Guide](https://docs.aws.amazon.com/sdk-for-javascript/v3/developer-guide) and [API reference](https://docs.aws.amazon.com/AWSJavaScriptSDK/v3/latest)
- [X] I've checked [AWS Forums](https://forums.aws.amazon.com) and [StackOverflow](https://stackoverflow.com/questions/tagged/aws-sdk-js).
- [X] I've searched for [previous similar issues](https://github.com/aws/aws-sdk-js-v3/issues) and didn't find any solution.

### Describe the bug

DynamoDBDocument update method was failing when providing an UpdateItem request that had 'AttributeUpdates' that contains objects.

The error received is as follows:

**TypeError: Cannot read properties of undefined (reading '0')
    at AttributeValue.visit (/var/runtime/node_modules/@aws-sdk/client-dynamodb/dist-cjs/models/models_0.js:645:40)
    at se_AttributeValue (/var/runtime/node_modules/@aws-sdk/client-dynamodb/dist-cjs/protocols/Aws_json1_0.js:2948:38)
    at Value (/var/runtime/node_modules/@aws-sdk/client-dynamodb/dist-cjs/protocols/Aws_json1_0.js:2972:23)
    at applyInstruction (/var/runtime/node_modules/@aws-sdk/smithy-client/dist-cjs/object-mapping.js:73:33)
    at take (/var/runtime/node_modules/@aws-sdk/smithy-client/dist-cjs/object-mapping.js:44:9)
    at se_AttributeValueUpdate (/var/runtime/node_modules/@aws-sdk/client-dynamodb/dist-cjs/protocols/Aws_json1_0.js:2970:37)
    at /var/runtime/node_modules/@aws-sdk/client-dynamodb/dist-cjs/protocols/Aws_json1_0.js:2943:20
    at Array.reduce (<anonymous>)
    at se_AttributeUpdates (/var/runtime/node_modules/@aws-sdk/client-dynamodb/dist-cjs/protocols/Aws_json1_0.js:2939:34)
    at AttributeUpdates (/var/runtime/node_modules/@aws-sdk/client-dynamodb/dist-cjs/protocols/Aws_json1_0.js:3516:34)**

We did read through https://github.com/aws/aws-sdk-js-v3/issues/3849 and https://github.com/aws/aws-sdk-js-v3/issues/5360 but neither of these have a satisfactory solution. 

At first, we tried providing the update method unmarshalled input like follows:
```{
    "TableName": "FlyhtMap-ai-dev-AircraftState",
    "Key": {
        "companyCode": "AKT",
        "aircraftReg": "C-FTIZ"
    },
    "AttributeUpdates": {
        "outTime": {
            "Action": "PUT",
            "Value": {
                "utc": "2023-11-13T19:28:26Z",
                "isEstimated": false
            }
        },
        "offTime": {
            "Action": "PUT",
            "Value": {
                "utc": "2023-11-13T19:37:04Z",
                "isEstimated": false
            }
        },
        "onTime": {
            "Action": "PUT",
            "Value": {
                "utc": null,
                "isEstimated": false
            }
        },
        "inTime": {
            "Action": "PUT",
            "Value": {
                "utc": null,
                "isEstimated": false
            }
        }
    }
}
```

But we received the error provided above. 
Then we added a middleware function to marshall the objects ourselves after reading some stackOverflow posts. 

```
client.middlewareStack.add(
      (next, context) => async (args) => {
          let updateArgs = JSON.parse(JSON.stringify(args));
          const attributeUpdates = updateArgs?.input?.AttributeUpdates;
          if (context?.commandName === 'UpdateItemCommand') {
              for (const property in attributeUpdates) {
                  if(Object.keys(attributeUpdates[property].Value).length > 1){
                          let valueToUpdate = attributeUpdates[property].Value;
                           if (typeof (valueToUpdate) === 'object') {
                              const hasAssignedType = valueToUpdate.hasOwnProperty('M');
                              if (!hasAssignedType) {
                                  valueToUpdate = marshall(valueToUpdate);
                                  const value = {
                                      M: valueToUpdate
                                  };
                                  valueToUpdate = value;
                              }
                          }
                          attributeUpdates[property].Value = valueToUpdate;
                  }
              }
          }
          console.log("Updated Object: %j", updateArgs);
          const result = await next(updateArgs);
          return result;
      },
      {
          step: "serialize",
          name: "improveMarshalling2"
      }
  );
```

With this middleware function, the object passed forward looks like:
```
{
    "middlewareStack": {},
    "input": {
        "TableName": "FlyhtMap-ai-dev-AircraftState",
        "Key": {
            "companyCode": "AKT",
            "aircraftReg": "C-FTIZ"
        },
        "AttributeUpdates": {
            "outTime": {
                "Action": "PUT",
                "Value": {
                    "M": {
                        "utc": {
                            "S": "2023-11-13T19:28:26Z"
                        },
                        "isEstimated": {
                            "BOOL": false
                        }
                    }
                }
            },
            "offTime": {
                "Action": "PUT",
                "Value": {
                    "M": {
                        "utc": {
                            "S": "2023-11-13T19:37:04Z"
                        },
                        "isEstimated": {
                            "BOOL": false
                        }
                    }
                }
            },
            "onTime": {
                "Action": "PUT",
                "Value": {
                    "M": {
                        "utc": {
                            "NULL": true
                        },
                        "isEstimated": {
                            "BOOL": false
                        }
                    }
                }
            },
            "inTime": {
                "Action": "PUT",
                "Value": {
                    "M": {
                        "utc": {
                            "NULL": true
                        },
                        "isEstimated": {
                            "BOOL": false
                        }
                    }
                }
            }
        }
    }
}
```

Which resulted in _the same error_.

We also tried both approaches with a flat input and, in that case, both approaches worked (there was no error, the database updated). 

This code runs in a lambda using the Node18x runtime. I was unable to find a way to obtain the exact version of the SDK being used in said runtime.
From the lambda logs:
INIT_START Runtime Version: nodejs:18.v18	Runtime Version ARN: arn:aws:lambda:us-west-2::runtime:d949ec0248c1b8fbc8cbc6df1c986e05fb933506b51d53be82d46f9a37125bea

### SDK version number

@aws-sdk/client-dynamodb@3.362.0

### Which JavaScript Runtime is this issue in?

Node.js

### Details of the browser/Node.js/ReactNative version

Node 18.18 - again lambda runtime I don't have visibility into this

### Reproduction Steps

```
//Lambda code - you will need a dynamo table.

const lambdaUtils = require('./lambdaUtils');
const productUtil = require(process.env.PRODUCT_LAYER_PATH + 'util');
const { DynamoDBClient, UpdateItemCommand } = require('@aws-sdk/client-dynamodb');
const { DynamoDBDocument } = require('@aws-sdk/lib-dynamodb');
const { marshall } = require("@aws-sdk/util-dynamodb");
const client = new DynamoDBClient({});
const awsSDK3 = DynamoDBDocument.from(client);

client.middlewareStack.add(
    (next, context) => async (args) => {
        let updateArgs = JSON.parse(JSON.stringify(args));
        const attributeUpdates = updateArgs?.input?.AttributeUpdates;
        if (context?.commandName === 'UpdateItemCommand') {
            
            for (const property in attributeUpdates) {
                if(Object.keys(attributeUpdates[property].Value).length > 1){
                        let valueToUpdate = attributeUpdates[property].Value;

                         if (typeof (valueToUpdate) === 'object') {
                            const hasAssignedType = valueToUpdate.hasOwnProperty('M');
                            if (!hasAssignedType) {
                                valueToUpdate = marshall(valueToUpdate);
                                const value = {
                                    M: valueToUpdate
                                };
                                valueToUpdate = value;
                            }
                        }

                        attributeUpdates[property].Value = valueToUpdate;
                }
                
                
            }
        }
        
        console.log("Updated Object: %j", updateArgs);
        const result = await next(updateArgs);
        return result;
    },
    {
        step: "serialize",
        name: "improveMarshalling2"
    }
);

//Our dynamo table name is an environment variable
const AIRCRAFT_STATE_TABLE = process.env.DDB_AIRCRAFT_STATE_TABLE_NAME;

exports.handler = async (event) => {
    try {
        const requestBody = event.body;
        
        const message = JSON.parse(requestBody);
        
        await updateIgnoreNulls(AIRCRAFT_STATE_TABLE, { "companyCode": message.companyCode, "aircraftReg": message.aircraftReg }, message);
        
        let response = productUtil.httpResponse(200, "Success");
        console.log("response: %j", response);

        return response;
    } catch (error) {
        console.error(error);
        return productUtil.httpResponse(500, JSON.stringify(error));
    }

};

async function updateIgnoreNulls(tableName, key, record) {
    record.updatedAt = (new Date().toISOString());

    let params = {
        TableName: tableName,
        Key: key,
        AttributeUpdates: skipNullAttributes(key, record)
    };

    console.log("Update Statement: %j", params);
    return await awsSDK3.update(params);
}

function skipNullAttributes(key, obj) {
    let attributes = {};

    for (let property in obj) {
        if (obj[property] && !key[property] && typeof(obj[property]) === 'object') {
            attributes[property] = { Action: 'PUT', Value: obj[property]};
        }
    }
    return attributes;
}


//Example event:
/*
{
  "resource": "/",
  "path": "/",
  "httpMethod": "POST",
  "headers": {
    "Accept-Encoding": "gzip,deflate",
    "Content-Type": "application/json; charset=UTF-8",
    "Host": "zb0fw1yxwf.execute-api.us-west-2.amazonaws.com",
    "User-Agent": "Apache-HttpClient/4.3.5 (java 1.5)",
    "X-Amzn-Trace-Id": "Root=1-654c80ff-02d9071611e52dcf1988efaf",
    "X-Forwarded-For": "208.118.96.82",
    "X-Forwarded-Port": "443",
    "X-Forwarded-Proto": "https"
  },
  "multiValueHeaders": {
    "Accept-Encoding": [
      "gzip,deflate"
    ],
    "Content-Type": [
      "application/json; charset=UTF-8"
    ],
    "Host": [
      "zb0fw1yxwf.execute-api.us-west-2.amazonaws.com"
    ],
    "User-Agent": [
      "Apache-HttpClient/4.3.5 (java 1.5)"
    ],
    "X-Amzn-Trace-Id": [
      "Root=1-654c80ff-02d9071611e52dcf1988efaf"
    ],
    "X-Forwarded-For": [
      "208.118.96.82"
    ],
    "X-Forwarded-Port": [
      "443"
    ],
    "X-Forwarded-Proto": [
      "https"
    ]
  },
  "queryStringParameters": null,
  "multiValueQueryStringParameters": null,
  "pathParameters": null,
  "stageVariables": null,
  "requestContext": {
    "resourceId": "d4v0mg7ttl",
    "resourcePath": "/",
    "httpMethod": "POST",
    "extendedRequestId": "OHkYBHqvvHcEJJg=",
    "requestTime": "09/Nov/2023:06:49:35 +0000",
    "path": "/ai-dev",
    "accountId": "184358129819",
    "protocol": "HTTP/1.1",
    "stage": "ai-dev",
    "domainPrefix": "zb0fw1yxwf",
    "requestTimeEpoch": 1699512575899,
    "requestId": "a10eea40-487f-488b-8c90-89ce138896a1",
    "identity": {
      "cognitoIdentityPoolId": null,
      "accountId": null,
      "cognitoIdentityId": null,
      "caller": null,
      "sourceIp": "208.118.96.82",
      "principalOrgId": null,
      "accessKey": null,
      "cognitoAuthenticationType": null,
      "cognitoAuthenticationProvider": null,
      "userArn": null,
      "userAgent": "Apache-HttpClient/4.3.5 (java 1.5)",
      "user": null
    },
    "domainName": "zb0fw1yxwf.execute-api.us-west-2.amazonaws.com",
    "apiId": "zb0fw1yxwf"
  },
  "body": "{\n \"companyCode\": \"XLR\",\n \"aircraftReg\": \"A9C-SJW\",\n \"outTime\": {\n\t  \"utc\": \"2023-11-09T05:02:47Z\",\n\t  \"isEstimated\": false\n\t},\n \"offTime\": {\n\t  \"utc\": \"2023-11-09T05:20:21Z\",\n\t  \"isEstimated\": false\n\t},\n \"onTime\": {\n\t  \"utc\": null,\n\t  \"isEstimated\": false\n\t},\n \"inTime\": {\n\t  \"utc\": null,\n\t  \"isEstimated\": true\n\t}}\n",
  "isBase64Encoded": false
}
*/

```

### Observed Behavior

Cloudwatch logs: 
[log-events-viewer-result.csv](https://github.com/aws/aws-sdk-js-v3/files/13343074/log-events-viewer-result.csv)

### Expected Behavior

We expected the library to either marshal the input correctly itself or to accept the correctly marshaled input we provided with the middleware stack function.

**Comment by aBurmeseDev** [comment]
Hi @third-party-integrations-flyht - thanks for reaching out. 

Upon my research, I've come across the similar issues linked below, including the ones that you mentioned. Most, if not all, of them have been fixed with https://github.com/aws/aws-sdk-js-v3/pull/5365 in [v3.429.0](https://github.com/aws/aws-sdk-js-v3/releases/tag/v3.429.0). Can you confirm the version you're on just so we can rule it out?

https://github.com/aws/aws-sdk-js-v3/issues/3849
https://github.com/aws/aws-sdk-js-v3/issues/4155
https://github.com/aws/aws-sdk-js-v3/issues/4703
https://github.com/aws/aws-sdk-js-v3/issues/5357
https://github.com/aws/aws-sdk-js-v3/issues/5360

**Comment by third-party-integrations-flyht** [comment]
Hi @aBurmeseDev, thank you for your prompt response!
That is probably one of the reasons why we couldn't debug the issue in full, as we do not explicitly install the aws-sdk modules in our Lambda/Lambda Layers code, so it is not in our package.json files.
There used to be a way to check the SDK version in SDK v2, by pulling the `AWS.VERSION` from the environment variables at the runtime, but since the SDK v3 is now modularized how can we target `"@aws-sdk/client-dynamodb"` or `"@aws-sdk/util-dynamodb"` directly?

**Comment by third-party-integrations-flyht** [comment]
Hi again!
Following up on our previous comment, the issue seems to be mitigated currently: we managed to control the version by providing package.json with the aws-sdk individual modules in the dependencies section.
So all and all the issue has been fixed by explicitly providing v3.429.0 (as the fix version mentioned above).
However, this leaves us with one question, if the version was not controlled, assuming we were on 3.362.0 for the past little while: how come it affected us only at the end of October?

**Comment by aBurmeseDev** [comment]
Hi @third-party-integrations-flyht - apologies for the delay. It sounds like we were able to confirm this was fixed in the [v3.429.0](https://github.com/aws/aws-sdk-js-v3/releases/tag/v3.429.0) which is awesome! The fix was merged and released on Oct 16 but there might be many reason why you didn't run into this till later. My guess would be that one or more of the column level values became undefined. 

If you run into any other issues, please feel free to reach out anytime!
Best,
John
